### DISCUSSION TITLE
Comparison of solvers

### DISCUSSION THREAD

**Discussion by ArjunM21** [yalmip/YALMIP#1060] [Optimization modelling] [answered]
Hey all.

I wrote an optimization code with the help I got from here and ran it successfully for multiple cases. Now I want to do a comparison of various solvers of YALMIP and know how fast they can solve the problem that I have. My problem can be solved using Branch and Bound and similar methods. Currently, YALMIP is currently using sedumi to solve the problem. Kindly help me in choosing various solvers for the same. 

This is needed for a bigger problem whereby I want to know if the constraints I have given are being run/considered because ideally the number of nodes that the solver goes through should have been reduced, which has not happened. All the commented statements are the constraints I want to be able to use to reduce the number of nodes.

Please find the code below.

I am sorry if my understanding of this is not good enough as I am still figuring things out. 
Thank you.
```

clc;
clear all;
close all;
yalmip('clear');
C = [1 0 0 ; 1 1 -1; 1 1 -1; 1 1 -1; 0 1 0; 1 0 -1; 0 1 -1; 0 0 1]; %1,5,8

t = sdpvar(8);
q = binvar(8,1);
Q = diag(q);
q0 = [0;0;0;0;0;1;1;1];
Q0 = diag(q0);

W1 = [1 0 0;0 0 0;0 0 0];
W2 = [1 1 -1;1 1 -1;-1 -1 1];
W3 = [1 1 -1;1 1 -1;-1 -1 1];
W4 = [1 1 -1;1 1 -1;-1 -1 1];
W5 = [0 0 0;0 1 0;0 0 0];
W6 = [1 0 -1;0 0 0;-1 0 1];
W7 = [0 0 0;0 1 -1;0 -1 1];
W8 = [0 0 0;0 0 0;0 0 1];
W = {W1, W2, W3, W4, W5, W6, W7, W8};

Constraints = [];
Constraints = [sum(q) == 3]; 
ch = 100;
tol = 1e-2;

Gdif = [];

while(ch > tol)
    Sum1 = [W{:}]*kron(q0,eye(3));
    disp(Sum1);
    Sum2 = [W{:}]*kron(ones(8,1),eye(3));
    G0 = trace(C*inv(Sum1)*C');
    for i = 1:8
        Gdif = [Gdif, -trace(inv(Sum1)*value(W{i})*inv(Sum1)*(Sum2))];
    end 
    Constraints = [Constraints, G0 + (Gdif *(q-q0)) <= t];
    Constraints = [Constraints, [t C;C' C'*Q*C] >=0];
    Constraints = [Constraints, t>=0];
%     Constraints = [Constraints, value(q0(2))+value(q0(3))+value(q0(4))<=1];
%     Constraints = [Constraints, value(q0(1))+value(q0(6))+value(q0(8))<=2];
%     Constraints = [Constraints, value(q0(5))+value(q0(7))+value(q0(8))<=2];
%     Constraints = [Constraints, value(q0(1))+value(q0(7))<=1];
%     Constraints = [Constraints, value(q0(5))+value(q0(6))<=1];
    options = sdpsettings('solver','bnb');
    ss = optimize(Constraints, trace(t), options);
    q0 = round(value(q));
    Q0 = diag(q0);
    t0 = round(value(trace(t))); 
    disp(G0);
    disp(t0);
    ch = abs(t0 - trace(C/(C'*Q0*C)*C'));
    Gdif = [];
end 
```

**Comment by johanlofberg** [comment, marked as answer]
YALMIP selecting sedumi means you don't have mosek installed, which you should have as it is the best SDP solver.

Your termination criteria looks strange, as you have a variable called ch which one would assume means change, but the variable does not compare current and previous objective, but is computed completely from the current solution.

The constraint` Constraints = [Constraints, t>=0]; ` can be removed as you already have that in the model

and the commented lines makes absolutely no sense as you are referencing a constant q0, and it makes even less sense as you apply value on a constant

there was also a bug in bnb which showed up when using mosek here, so when switching to mosek you have to install the develop branch.

**Reply by johanlofberg** [reply]
and I would not trust `G0 + (Gdif *(q-q0)) <= t` as left side is a scalar and right side is a matrix so the whole thing is a semidefinite constraint, i.e. my guess is you don't really know what you have there

**Reply by ArjunM21** [reply]
Thank you for the reply. Actually, I am still trying to make sense of the whole thing myself, so please bear with me.

The ch variable was used to compare the value of t0 and the value of the trace of the matrix multiplication. How can I compare the present and previous t0 values if I am updating them in each loop? 

I did not know about mosek, so I will install it and see.

Let me try to explain what I am trying to do with the commented section. The second, third, and fourth values of q0 should add up to 1. That is the condition I want to implement. Ideally adding the commented lines should help the solver go through fewer nodes. But I do not understand how the solver is solving here. I need some help with this part.

`G0 + (Gdif *(q-q0)) <= trace(t)` is how its supposed to be.

Also, why do you think my code runs properly and I am getting expected results?

**Reply by johanlofberg** [reply]
We don't know what it is supposed to be. My guess is you are linearizing something like trace(C * (C' * Q * C)^-1 * C) or something like that, and then it makes even less sense that you include both the linearization, and the linear SDP form of that?

q0 is constant, so to say that it should sum up to some value makes no sense. You don't seem to understand the difference between a constant (such as q0) and a decision variable (q) which has a value after you have solved a problem (value(q))

A standard loop to detect small change in optimal t would be

```
ch = inf;
while(ch > tol)
...
newt0 = round(value(trace(t)));
ch = newt0-t0;
t0 = newt0;
end
```

**Reply by ArjunM21** [reply]
Thank you for the input. 

I have given the SDP condition because I was told that it would ensure only necessary values of q would be calculated. The linearization part is what I am testing out. 

I tried installing mosek and ran my code, which gave me NaN as the result, so I am sure there is something wrong. What did you mean by installing the develop branch? I do not understand the code properly yet myself, but I am trying to make sense of it all. The paper from which I am trying this out is right, so it has to be a mistake on my part. The first constraint is the linearization and the second constraint is for making sure certain values of q are not considered.

In regards to the ch computation, I need to find the difference between the value of trace(t) and the value of G0 for the particular value of q found out. This difference has to be within tolerance for the loop to break.

Can you tell me all the reasons you think the code below does not make sense to you, so that I can find answers to them individually and correct them? 

```
clc;
clear all;
close all;
yalmip('clear');
C = [1 0 0 ; 1 1 -1; 1 1 -1; 1 1 -1; 0 1 0; 1 0 -1; 0 1 -1; 0 0 1]; %1,5,8
% C = [1 0 0 ; 0 1 0; 0 1 0; 0 1 0; 0 0 1; 0 1 -1;-1 1 0; 1 -1 1]; %1,2,5

t = sdpvar(8);
q = binvar(8,1);
Q = diag(q);
q0 = [0;0;0;0;0;1;1;1];
Q0 = diag(q0);

W1 = [1 0 0;0 0 0;0 0 0];
W2 = [1 1 -1;1 1 -1;-1 -1 1];
W3 = [1 1 -1;1 1 -1;-1 -1 1];
W4 = [1 1 -1;1 1 -1;-1 -1 1];
W5 = [0 0 0;0 1 0;0 0 0];
W6 = [1 0 -1;0 0 0;-1 0 1];
W7 = [0 0 0;0 1 -1;0 -1 1];
W8 = [0 0 0;0 0 0;0 0 1];
W = {W1, W2, W3, W4, W5, W6, W7, W8};

Constraints = [];
Constraints = [sum(q) == 3]; 
ch = 100;
tol = 1e-2;

Gdif = [];

while(ch > tol)
    Sum1 = [W{:}]*kron(q0,eye(3));
    disp(Sum1);
    Sum2 = [W{:}]*kron(ones(8,1),eye(3));
    G0 = trace(C*inv(Sum1)*C');
    for i = 1:8
        Gdif = [Gdif, -trace(inv(Sum1)*value(W{i})*inv(Sum1)*(Sum2))];
    end 
    disp('c')
    Constraints = [Constraints, G0 + (Gdif *(q-q0)) <= trace(t)];
    Constraints = [Constraints, [t C;C' C'*Q*C] >=0];
    Constraints = [Constraints, trace(t)>=0];
    ss = optimize(Constraints, trace(t));
    q0 = round(value(q));
    Q0 = diag(q0);
    t0 = round(value(trace(t))); 
    disp(G0);
    disp(t0);
    ch = abs(t0 - trace(C*inv(Sum1)*C'));
    Gdif = [];
end 
```

**Reply by johanlofberg** [reply]
regarding Mosek: Have you solved a trivially mall problem to see that it actually solves it, and if not have you turned on 'debug' so that you see the error message if there is one. My guess is that you've failed to setup license etc correctly

This

`  Constraints = [Constraints, trace(t)>=0];`

is obviously redundant when you include the SDP constraint above it

Regarding the general question here, it is really only going in circles and I can only advice you to take that discussion with your supervisor, as it really isn't a YALMIP issue.  The fact that the use of both the exact necessary and sufficient SDP constraint and the outer approximation cuts at the same time is weird, I will only describe with an analogous problem

```
% We want to minimize sum(x) with x'*x<=1
% We do not want to use quadratics, so we investigate
% an SDP form vs an outer approximation strategy

% Just use the SDP form
x=sdpvar(2,1);
optimize([1 x';x eye(2)]>=0,sum(x))

% Use tangent cuts iteratively to build
% outer approimation of feasible set
Model = [-1<=x<=1];
for i = 1:10
    plot(Model);drawnow
    optimize(Model,sum(x));
    x0 = value(x)
    Model = [Model, x0'*x0 + 2*x0'*(x-x0) <= 1];
end

% Makes no sense to include the SDP
% and then build the outer approximation
% since the outer approximation cuts are redundant
Model = [-1<=x<=1,[1 x';x eye(2)]>=0];
for i = 1:10
    plot(Model);drawnow
    optimize(Model,sum(x));
    x0 = value(x)
    Model = [Model, x0'*x0 + 2*x0'*(x-x0) <= 1];
end

```

**Reply by ArjunM21** [reply]
Hey, thank you so much for your reply. 
I did try running a sample problem and it ran perfectly. But when I try running my code, again the output is NaN. But it runs smoothly when I use SeDuMi-1.3. I am not sure what's wrong.

The constraint that you mentioned is necessary cus the previous constraint only takes into consideration the semi-definiteness of matrix t, which is one condition and the next line I thought took care of all diagonal values of the t matrix are positive or 0. Also when I try to remove that condition and run this, the answer is again NaN.

I understand what you are saying, but when I try it like that, I am not getting any solution. Do you think I should rewrite this whole thing again? Also why am I getting the right answer using Sedumi and not Mosek? That part is still not clear to me.

**Reply by johanlofberg** [reply]
The constraint t>=0 is redundant when you include  [t C;C' C'*Q*C] >=0, and the iteratively generated cuts G0 + (Gdif *(q-q0)) <= trace(t) should be redundant (assuming you are doing things correctly, which I honestly not would bet much on) if you include [t C;C' C'*Q*C] >=0, since the whole idea is to iteratively outer-approximate the original nonlinear constraint which is equivalent to [t C;C' C'*Q*C] >=0

The answer is Nan, well then what does the solver say in the diagnostics? When did you get Nan?
